Autowiki Stock Parts Template

Peliex opened the pull request:
## About The Pull Request

Autowiki functionality for stock parts

stockparts.dm extends the autowiki datum, implementing multiple functions to upload images of and parse information about all the craftable stock parts in the game a player is likely to encounter.

The template can be called on the wiki by using `{{Autowiki/Content/StockParts}}`

[VIEW OUTPUT HERE](https://tgstation13.org/wiki/Template:Autowiki/Content/StockParts)

## Why It's Good For The Game

Better documentation means lower learning curve, items that are properly up-to-date is a good start.
This will probably branch into a few other autowiki additions for the other various items from research, has the potential to lead to a full documentation of the autolathe if done correctly.
## Changelog
:cl:Senefi
code: Autowiki module for stock parts
fix: Emergency lights no longer runtime when created in nullspace
/:cl:

Peliex committed:
initial commit

Implemented autowiki functionality for creating a list of all stock parts that can be researched and crafted

Peliex committed:
changed switch types for comparison

Peliex committed:
Merge remote-tracking branch 'upstream/master' into wikiparts

Peliex committed:
Merge branch 'master' into wikiparts

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+		var/list/entry_contents = list()
+
+		// Populate entry_contents list
+		entry_contents["name"]        = escape_value(format_text(stock_part.name))
+		entry_contents["icon"]        = escape_value(format_text(create_icon(stock_part)))
+		entry_contents["desc"]        = escape_value(format_text(stock_part.desc))
+		entry_contents["type"]        = escape_value(format_text("[stock_part.type]"))
+		entry_contents["id"]          = escape_value(format_text(recipe.id))
+		entry_contents["tier"]        = escape_value(format_text("[stock_part.rating]"))
+		entry_contents["sources"]     = escape_value(format_text(generate_source_list(recipe)))
+		entry_contents["node"]        = escape_value(format_text(required_node.display_name))
+		entry_contents["materials"]   = escape_value(format_text(generate_material_list(recipe)))
```
No tab alignment

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+		entry_contents["name"]        = escape_value(format_text(stock_part.name))
+		entry_contents["icon"]        = escape_value(format_text(create_icon(stock_part)))
+		entry_contents["desc"]        = escape_value(format_text(stock_part.desc))
+		entry_contents["type"]        = escape_value(format_text("[stock_part.type]"))
```
Don't include this, it's not useful for players

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+		// contains current loop iteration's template parameters in key value pairs
+		var/list/entry_contents = list()
+
+		// Populate entry_contents list
```
Useless comment

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+				else
+					continue
+
+		// Instantiate stock part
```
Useless comment

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+	// Find all valid stock parts and their references
+	for(var/part_type in subtypesof(/obj/item/stock_parts))
+		// Skip subspace subtype as it is a grouping subtype and invalid
+		if(part_type == /obj/item/stock_parts/subspace)
```
This should be done through some `abstract_type` system instead

Peliex replied:
I'm looking into the typecache system so that way I can finally finish this project up, but I'm not sure what you mean by `abstract_type`

Mothblocks replied:
We have a trick where we specify something like this on types:

```dm
/datum/category
    var/abstract_type = /datum/category

/datum/category/item
    // use as normal
```

...and when iterating, we skip over everything that has `abstract_type == type`

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+		// All other battery types could possibly be put in another list?
+		// As it is, power cells are barely stock parts with how much extra functionality they have
+		if(ispath(part_type, /obj/item/stock_parts/cell))
+			switch(part_type)
+				if(/obj/item/stock_parts/cell)
+				if(/obj/item/stock_parts/cell/high)
+				if(/obj/item/stock_parts/cell/super)
+				if(/obj/item/stock_parts/cell/hyper)
+				if(/obj/item/stock_parts/cell/bluespace)
+				else
+					continue
```
Make this a typecache and explain why something would be in here

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+// Autowiki Stock Parts
+/datum/autowiki/stockparts
```
Nit: `stock_parts`

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+// Autowiki Stock Parts
```
Use `///` and better explain what this is creating exactly

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+	var/filename = SANITIZE_FILENAME(escape_value(stock_part.icon_state))
+	upload_icon(icon(stock_part.icon, stock_part.icon_state, SOUTH, 1, FALSE), filename)
+
+	return "Autowiki-" + filename + ".png"
```
Use `[]` for string interpolation

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+	// Check for Autolathe bitflag and add to list, optionally add comma if not first
+	if((recipe.build_type & AUTOLATHE) == AUTOLATHE)
+		if(source_list)
+			source_list += ", "
```
Instead of this, create a list and call `.Join(", ")` in the end

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+		// Append material cost to string list
+		material_list += "[recipe.materials[ingredient_type]] "
+
+		// Append material name to string list
```
You comment way too much, comments should explain why and not what--this is all just noise

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,127 @@
+	for(var/ingredient_type in recipe.materials)
+		var/datum/material/ingredient = new ingredient_type()
+		// If we aren't the first entry, add a line break
+		if(!initial)
+			material_list += "<br>"
```
Kill `initial` entirely and instead use a list with Join("<br>")

Peliex committed:
Changed datum subtype name from stockparts to stock_parts

Peliex committed:
Stripped code comments

Most of them were useful when implementing but serve little purpose now but adding lines

Peliex committed:
removed tab spacing

Peliex committed:
removed id content entry, changed generate_material_list to use a list to generate recipes instead of appending strings

Mothblocks commented:
Unstaled as I did not respond to feedback

Peliex committed:
Implemented abstract type and only output results with a valid recipe

Peliex committed:
Merge remote-tracking branch 'upstream/master' into wikiparts

Peliex committed:
fix emergency light runtime when instantiated in nullspace

Peliex committed:
Reimplemented battery check to remove empty batteries from output file

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -2,30 +2,28 @@
 	page = "Template:Autowiki/Content/StockParts"
 
+	var/list/battery_whitelist = list(
+	/obj/item/stock_parts/cell,
+	/obj/item/stock_parts/cell/high,
+	/obj/item/stock_parts/cell/super,
+	/obj/item/stock_parts/cell/hyper,
+	/obj/item/stock_parts/cell/bluespace)
```
Indent once, end parentheses on its own line, trailing comma

https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#multi-lining

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+/datum/autowiki/stock_parts
+	page = "Template:Autowiki/Content/StockParts"
+
+	var/list/battery_whitelist = list(
```
Looks like this is completely unused?

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+	var/filename = SANITIZE_FILENAME(escape_value(stock_part.icon_state))
+	upload_icon(icon(stock_part.icon, stock_part.icon_state, SOUTH, 1, FALSE), filename)
+
+	return "Autowiki-" + "[filename]" + ".png"
```
```suggestion
	return "Autowiki-[filename].png"
```

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+/datum/autowiki/stock_parts/proc/generate_source_list(datum/design/recipe)
+	var/source_list = ""
+
+	if((recipe.build_type & PROTOLATHE) == PROTOLATHE)
```
Why bother checking `== PROTOLATHE`? You only need this kind of pattern when dealing with multiple bitflags

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+
+	if((recipe.build_type & AUTOLATHE) == AUTOLATHE)
+		if(source_list)
+			source_list += ", "
```
You resolved my comment but did not do it. Use Join instead.

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+	return source_list
+
+/datum/autowiki/stock_parts/proc/generate_material_list(datum/design/recipe)
+
```
```suggestion
```

Mothblocks commented on code/modules/autowiki/pages/stockparts.dm:
```diff
@@ -0,0 +1,92 @@
+	for(var/ingredient_type in recipe.materials)
+		var/datum/material/ingredient = new ingredient_type()
+
+		materials.Add("[recipe.materials[ingredient_type]] " + "[ingredient.name]")
```
```suggestion
		materials += "[recipe.materials[ingredient_type]] [ingredient.name]"
```

Mothblocks commented on code/modules/power/cell.dm:
```diff
@@ -462,8 +462,9 @@
 	var/area/A = get_area(src)
-	if(!A.lightswitch || !A.light_power)
-		charge = 0 //For naturally depowered areas, we start with no power
+	if(A)
```
Can you change this to just `area` while you are here

Mothblocks commented on code/modules/research/stock_parts.dm:
```diff
@@ -246,6 +246,7 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
 	///Used when a base part has a different name to higher tiers of part. For example, machine frames want any servo and not just a micro-servo.
 	var/base_name
 	var/energy_rating = 1
+	var/abstract_type = /obj/item/stock_parts
```
Document

Peliex committed:
Merge remote-tracking branch 'upstream/master' into wikiparts

Peliex committed:
Implemented code review changes, thanks mothblocks

Peliex committed:
fixed broke shit

Peliex committed:
style my ass

Peliex committed:
Merge branch 'master' into wikiparts

Peliex committed:
Merge branch 'master' into wikiparts

Peliex committed:
Merge branch 'master' into wikiparts

Peliex committed:
Merge branch 'master' into wikiparts
